### PR TITLE
nixos/weechat: add module

### PR DESCRIPTION
**Pull request by Ma27**
###### Motivation for this change

Follow-up for #33523. Removed static uid/gid as proposed here (https://github.com/nixos/nixpkgs/pull/33523#issuecomment-372455264) and added support for loading custom scripts.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by Ma27** [comment]
@timokau after yesterday's discussion I was motivated to have a look at it. Seems as it's sufficient to load scripts by specifying their absolute paths (means, store paths :D). Would be cool if you could have a look at this.

@yegortimoshenko as I wanted to hack on a weechat module for NixOS I used your PR from January as base. It would be awesome if you could have a look at this! :)

File: nixos/modules/services/misc/weechat.nix
Comment (timokau) [inline nixos/modules/services/misc/weechat.nix]: So `name` is actually a path?

File: nixos/modules/services/misc/weechat.nix
Comment (timokau) [inline nixos/modules/services/misc/weechat.nix]: Why not just `"${pkgs.weechat-xmpp}/share/jabber.py"`? Or maybe even force the script packages to be in a standard format and only require the package.

File: nixos/modules/services/misc/weechat.nix
Comment (timokau) [inline nixos/modules/services/misc/weechat.nix]: I think it is better to use normal quotes here. I don't think the tex-like backtick/normal tick combination is very common in plain text.

File: nixos/modules/services/misc/weechat.nix
Comment (timokau) [inline nixos/modules/services/misc/weechat.nix]: Is this mkif really necessary? Wouldn't it just be an empty string in the case of length 0?

File: nixos/modules/services/misc/weechat.nix
Comment (timokau) [inline nixos/modules/services/misc/weechat.nix]: I don't have very much experience with the module system and merging and everything -- will these scripts be loaded *before* the rest of the init code is executed? I think they should be.

File: nixos/modules/services/misc/weechat.nix
Comment (timokau) [inline nixos/modules/services/misc/weechat.nix]: What does this do?

**Comment by timokau** [comment]
Thanks! Any reason you went with a nixos module? I feel like adding the configuration through a package wrapper may be nicer.

File: nixos/modules/services/misc/weechat.nix
Comment (Ma27) [reply nixos/modules/services/misc/weechat.nix]: You're absolutely right, this makes the expression way easier to read, thanks!

File: nixos/modules/services/misc/weechat.nix
Comment (Ma27) [reply nixos/modules/services/misc/weechat.nix]: I'm not sure either, but you're right, it to avoid breaking scripts it might be helpful to use `mkBefore` to be absolutely sure :)

File: nixos/modules/services/misc/weechat.nix
Comment (Ma27) [reply nixos/modules/services/misc/weechat.nix]: according to the docs (https://weechat.org/files/doc/stable/weechat_user.en.html#fifo_plugin), the `*` at the beginning of a string is needed when passing commands to weechat through `weechat_fifo`. This was actually introduced in the previous commit I didn't author, but it's more convenient to type `/cmd args...` without the star at the beginning.

File: nixos/modules/services/misc/weechat.nix
Comment (Ma27) [reply nixos/modules/services/misc/weechat.nix]: not sure though. I've seen this fairly often in the Nix(OS) context including descriptions for options and wanted to be convenient. But feel free to correct me, I'm not sure about this :)

File: nixos/modules/services/misc/weechat.nix
Comment (Ma27) [reply nixos/modules/services/misc/weechat.nix]: you're absolutely right! I originally wanted to define a pre-defined output structure for derivations which feature weechat scripts, but dropped the idea later and kept the (admittedly too complex submodule type).

File: nixos/modules/services/misc/weechat.nix
Comment (Ma27) [reply nixos/modules/services/misc/weechat.nix]: I was thinking about "name of the script", but well, there are only two hard things in computer science, right? ;)

But seriously, you're right, `path` is much better, however this can be dropped as you stated later on.

**Comment by Ma27** [comment]
> Any reason you went with a nixos module? I feel like adding the configuration through a package wrapper may be nicer.

I originally wanted to make the weechat module of @yegortimoshenko ready to merge and wanted to see if it's possible to easily add support for scripts (which is easily doable by adding `/script load $path` to `services.weechat.init`). Previously I assumed that scripts are only loadable if they're installed in `~/.weechat`, however absolute paths are possible as well.

I'm not sure if we want to duplicate the `weechat_fifo` the logic of the previously written module in the package (and if it's even possible), but I'm happy to discuss about it to see which is the best decision for nixpkgs :)

File: nixos/modules/services/misc/weechat.nix
Comment (timokau) [reply nixos/modules/services/misc/weechat.nix]: You're right, I can find a couple of examples of both simple quotes and `' quotes in `man configuration.nix`. I personally prefer simple quotes but since there is apparently no consensus I'm fine with whatever you prefer.

File: nixos/modules/services/misc/weechat.nix
Comment (Ma27) [reply nixos/modules/services/misc/weechat.nix]: I just used the code from the previous commit. However I think that this is the right way to go as the star is only used with `fifo`.
However the `/` at the beginning is also used when running weechat from an active session, so it's IMHO more convenient to keep this.

**Comment by Ma27** [comment]
I learned about `weechat_fifo` as an approach to "remotely control" weechat. You could even run a systemd unit during the runtime of `weechat` with new options.

However `--run-command` seems to be a reasonable alternative. Please give me some time to think about this, then we can continue the discussion:)

**Comment by Ma27** [comment]
btw, do you think we can sneak this into 18.09? :)

File: pkgs/applications/networking/irc/weechat/default.nix
Comment (timokau) [inline pkgs/applications/networking/irc/weechat/default.nix]: If subsequent `--run-command`s overwrite the first one, maybe we could check for `--run-command` or `-r` in `$@` and concat the two. It would add a bit of complexity but just silently overwriting may confuse users.

**Review by timokau — CHANGES_REQUESTED** [review]
Awesome, thanks! There is always a bit to nitpick, but looks good overall.

**Comment by timokau** [comment]
> 18.09

Not sure what the policy for the "feature freeze" is. I'd be in favor (since this doesn't update weechat and is unlikely to break anything) but we should ask one of the release managers once this is merged.

File: doc/package-notes.xml
Comment (Ma27) [reply doc/package-notes.xml]: unfortunately not. I tried it out with multiple `--run-command` arguments. I can try to merge both values though, but I'm not sure, how easy this will be...

File: doc/package-notes.xml
Comment (Ma27) [reply doc/package-notes.xml]: You can set values for your scripts before weechat actually starts. But you're right, re-reading it makes the  sentence quite odd :sweat_smile:

**Comment by Ma27** [comment]
@timokau thanks a lot for your feedback! Would you mind having a look at this a second time? :)

File: doc/package-notes.xml
Comment (timokau) [reply doc/package-notes.xml]: So basically just "you can configure WeeChat"? I think the formulation should be improved, at least I'm having a hard time with it.

File: doc/package-notes.xml
Comment (timokau) [reply doc/package-notes.xml]: Oh right, I forgot that plugins and scripts are two entirely different things. Maybe that should be made explicit in the docs. So this actually just adds all available plugins?

File: doc/package-notes.xml
Comment (timokau) [reply doc/package-notes.xml]: Alright

File: doc/package-notes.xml
Comment (timokau) [reply doc/package-notes.xml]: It *is* a bit of complexity (especially for a bash script), but I think it is worth it because the alternative can be very confusing. This should do the trick (assuming `init` is set correctly):

```bash
#!/bin/sh
init="" # init script

new_args=()
# remove --run-command arguments and remember their commands
while [ $# -gt 0 ]; do
	if [[ "$1" == "--run-command" ]] || [[ "$1" == "-r" ]]; then
		# subsequent `--run-commands` override the previous ones, as is default weechat behaviour
		commands_to_run="$2"
		shift
	else
		new_args+=("$1")
	fi
	shift
done

new_args+=("--run-command" "$init;$commands_to_run")

exec weechat "${new_args[@]}"

```

File: doc/package-notes.xml
Comment (timokau) [reply doc/package-notes.xml]: Yeah something like

```
scripts  = with pkgs.weechatScripts [
  xmpp
  matrix-protocol
]
```

Would be pretty neat. It would be possible to remain backwards compatibility, although I'm not sure if its worth it.

File: pkgs/applications/networking/irc/weechat/default.nix
Comment (timokau) [inline pkgs/applications/networking/irc/weechat/default.nix]: We probably shouldn't modify weechats standard behaviour of overriding previous `--run-commands` (that were explicitly passed by the user). Although I personally think aggregating them would make way more sense.

On the other hand weechat may change that in the future and there isn't really too much harm in aggregating them (who in their right mind would depend on the override behaviour?), so I'm not sure.

File: doc/package-notes.xml
Comment (timokau) [reply doc/package-notes.xml]: Oh I just saw that you already implemented that.

File: pkgs/applications/networking/irc/weechat/default.nix
Comment (timokau) [reply pkgs/applications/networking/irc/weechat/default.nix]: I've opened https://github.com/weechat/weechat/issues/1248 upstream for that.

File: doc/package-notes.xml
Comment (Ma27) [reply doc/package-notes.xml]: You're right, the terms `script` and `plugin` seem to be the same thing, I made the mistake several times as well.

However I don't think that it's needed to add another sentence here, when you have a look at the [WeeChat docs in the `nixpkgs` manual](https://github.com/NixOS/nixpkgs/blob/2eea92a1dc124e24744d57f53d9dbe10e85876e1/doc/package-notes.xml#L662), you'll see that how to configure and use `plugins` with the Nix package is explained in the previous paragraph.

File: doc/package-notes.xml
Comment (Ma27) [reply doc/package-notes.xml]: I'm not sure, but I'd even say that this is out of scope. This PR already includes script/config support for the package *and* a module and I'd say that splitting single scripts into their own metapackage should be done in a second PR.

Regarding the change itself: I'm definetely in favor of it, users don't need to know about the structure inside a given store path, they simply need to know their packages, furthermore this is way more readable IMHO.

Regarding backwards compatibility: I've seen several people using these packages, so I think that just moving them is a bad idea. Let's place them into `pkgs/top-level/aliases.nix` and print a deprecation warning if this attribute path is used.

File: doc/package-notes.xml
Comment (timokau) [reply doc/package-notes.xml]: Right, its easy to get lost in diffs and miss the context. However I think a short aside like "note that plugins are not the same as scripts. See the scripts section for more information" or something when plugins are first introduced may be useful.

Comment (timokau) [reply doc/package-notes.xml]: Also does that mean that no plugins are installed if I pass in a config without specifying plugins? The default (all plugins) should probably be kept if plugins aren't specified.

File: pkgs/applications/networking/irc/weechat/default.nix
Comment (timokau) [reply pkgs/applications/networking/irc/weechat/default.nix]: You currently append to `next` each time you find a `--run-command`, which is different from the default weechat behaviour.

However somebody upstream already said they'll "implement that quickly", so it should be fine as-is.

File: doc/package-notes.xml
Comment (timokau) [reply doc/package-notes.xml]: Scripts outside of the nix store would be an impurity that would make the config less portable. I see how it could be useful however, so if you want to support that its fine by me.

File: doc/package-notes.xml
Comment (timokau) [reply doc/package-notes.xml]: Why not change `inherit (config) plugins` to `plugins = config.plugins or availablePlugins`? Requiring that argument made sense when it was the only argument but makes less sense now. I suspect most people don't care much about a couple of MB in the closure size of an end user application.

File: doc/package-notes.xml
Comment (Ma27) [reply doc/package-notes.xml]: I guess you're right, packaging weechat scripts shouldn't be too hard.I'll see how much time it costs to implement this, then I may add this to this PR.

**Comment by Ma27** [comment]
@timokau I added myself as maintainer and introduced the `weechatScripts` metapackage to load scripts on startup. Would you mind having *another* look? :)

File: doc/package-notes.xml
Comment (timokau) [inline doc/package-notes.xml]: This part is out of date now right?

File: doc/package-notes.xml
Comment (timokau) [inline doc/package-notes.xml]: I would prefer those variables quoted, although there was some contoversy about that. So I won't insist on it.
Also applies to other files changed here.

File: pkgs/applications/networking/irc/weechat/default.nix
Comment (timokau) [inline pkgs/applications/networking/irc/weechat/default.nix]: A short comment that explains what this does would be nice.

File: pkgs/applications/networking/irc/weechat/default.nix
Comment (timokau) [inline pkgs/applications/networking/irc/weechat/default.nix]: This would double the `--run-commands` if https://github.com/weechat/weechat/issues/1248#issuecomment-419056035 is implemented right? If so, you should add a comment to the `version` attribute reminding whoever updates weechat to adjust that.

**Review by timokau — COMMENTED** [review]
Looks good, nearly done :)

File: doc/package-notes.xml
Comment (Ma27) [reply doc/package-notes.xml]: Admittedly I've never seen this pattern. Furthermore it's IIRC documented in Nixpkgs (or even Nix?), that `$out` exists as reference to the pending store path.

As you mention "some controversy about that", would you mind sending me the link to the discussion? I don't think that this is needed, but I'd like to see the arguments of both sides :)

Comment (timokau) [reply doc/package-notes.xml]: Yes, `$out` is a reference to the pending store path. I think as a shell variable it should still be quoted. I was referring to this discussion: https://github.com/NixOS/nixpkgs/pull/45657#discussion_r212835434

File: pkgs/applications/networking/irc/weechat/default.nix
Comment (timokau) [inline pkgs/applications/networking/irc/weechat/default.nix]: Now that this is [fixed upstream](https://github.com/weechat/weechat/issues/1248#event-1831302139), we could instead cherry-pick the upstream solution as a patch. Either that or add a comment to the `version` attribute to remind whoever updates weechat next to remove this.

Comment (Ma27) [reply pkgs/applications/networking/irc/weechat/default.nix]: now this was fast :D 
Unfortunately the patch doesn't apply onto v2.2, I'll try to rebase it, but it seems to be a bit of work 
as several translation files are affected.

Comment (timokau) [reply pkgs/applications/networking/irc/weechat/default.nix]: I was definitely positively surprised, nice when upstream is so reactive :)

Yeah no reason to rebase it I think. You could make use of fetchpatch's `includes` if the translation files are the only issues (that really are a *lot* of translation files). Otherwise the comment to the version is also fine.

Comment (Ma27) [reply pkgs/applications/networking/irc/weechat/default.nix]: ah dammit, `src/core/weechat.c` contains some conflicts as well. But I'll think about this tomorrow, I'll catch some sleep now :)

Comment (timokau) [reply pkgs/applications/networking/irc/weechat/default.nix]: No need to spend too much time on this then. We just need to make sure that the code doesn't stay around forever and for that the comment should be enough.

**Comment by timokau** [comment]
Thanks! Unfortunately ofBorg is reporting

```
undefined variable 'weechat' at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/eval-0-gleber.ewr1.nix.ci/nixos/modules/services/misc/weechat.nix:23:56
```

**Comment by timokau** [comment]
Great, thank you for your patience :) Looks good now.

Just FYI, GitHub doesn't send email notifications when you force-push. I only noticed because of ofBorgs rebuilds.

**Review by timokau — APPROVED** [review]

**Comment by timokau** [comment]
#46324 should also be included in 18.09 if we include this.

**Comment by samueldr** [comment]
Oof, the backport policies are really about updating software, and updating software revisions for non-major updates. I don't know if a new module should be backported. it's kind of a lot of things, and the package has deviated considerably compared to how it worked prior to this PR.

SInce nix, nixos and modules can be composed pretty easily, I'm not sure this should be backported. Though the *version update* (only) could and probably should be backported. (Though it does not seem to have security fixes, it does have stability fixes.)

I'm not strictly against backporting this, but I'm not sure it should be. I'm don'T know about previous precedents of backports of non-trivial module updates. And then this would create such a precedent.

Thoughts, @vcunat?

**Comment by timokau** [comment]
On 18-09-07 19:47, Samuel Dionne-Riel wrote:
> Oof, the backport policies are really about updating software, and updating software revisions for non-major updates. I don't know if a new module should be backported. it's kind of a lot of things, and the package has deviated considerably compared to how it worked prior to this PR.

Yes but the most "risky" changes are all additions, so no regressions. And since weechat is an end-user application and not a library, potential for breakage is limited and can be tested pretty quickly.

> SInce nix, nixos and modules can be composed pretty easily, I'm not sure this should be backported.

Can they? I haven't found a good way to compose modules from different NixOS versions yet. For me the most important part of this is the scripts support, which can be used by simply fetching the weechat package from unstable. So I don't care very strongly about backporting it, but I still think it would be better. I understand the arguments against it however -- feature freeze is feature freeze.

**Comment by timokau** [comment]
Note that while #46324 was a breakage, it was not a regression. Just a small bug in the newly added features.

**Comment by vcunat** [comment]
@samueldr: it's just a new module, right?  Those who don't enable it shouldn't be affected at all by that.

**Comment by timokau** [comment]
A new module and configurability added to the weechat package. Still shouldn't affect anybody not using it.

**Comment by samueldr** [comment]
Sorry, I meant to reply earlier.

I must have misread that it was changing an existing module. Probably by relying on the description "removing [...] adding support for [...]" instead of looking at the actual code.

In that case it is a fine candidate for a backport I guess.

**Comment by Ma27** [comment]
> Sorry, I meant to reply earlier.

No worries, we've learned at last NixCon that release management is quite a hard and time-consuming task :)

> In that case it is a fine candidate for a backport I guess.

great! I'm currently using the changes with a recent 18.09 checkout personally, I'll recheck and file a PR later to do the backport, ok?

**Comment by samueldr** [comment]
Yes, you can tag me in the backport PR. Thanks for making the PR!

**Comment by dtzWill** [comment]
Hey--how is this intended to be used? Anyone have an example?

The documentation suggests this is geared so that you can attach to the screen session (command to do so is mentioned, and modification to screenrc suggested, etc.), but that is a non-starter AFAICT since screen needs to be suid for that to work and our screen is not.

Which makes me wonder--given this was merged so recently... how are folks using it? Or is there a way to attach I'm missing (sorry!)? Or to run as a user service, perhaps? Thanks!

**Comment by Ma27** [comment]
Hey! Thanks a lot for you interest in this module!
Unfortunately I'll a bit busy in the next days, so please give me some time, then I'll have a deeper look at the issue, ok? :)

**Comment by dtzWill** [comment]
Sure thing! No hurry at all :).  See you in a few!

**Comment by Ma27** [comment]
first of all sorry for the delay, I meant write earlier.

After debugging I realised that I used a `screen` with `setuid` on my server (as I needed it for some experiments earlier). During next week I'll whats the best way to implement this and file a patch:)

**Pull request merged**
